src/nodes.ts

```diff
@@ -512,6 +541,204 @@ export class UnaryExistsOp extends Node {
   }
 }
 
+export class Spread extends Node {
+  readonly expression: Node;
+
+  constructor(
+    line: number,
+    column: number,
+    start: number,
+    end: number,
+    raw: string,
+    virtual: boolean,
+    expression: Node
+  ) {
+    super('Spread', line, column, start, end, raw, virtual);
+    this.expression = expression;
+  }
+}
+
+export class Range extends Node {
+  readonly left: Node;
+  readonly right: Node;
+  readonly isInclusive: boolean;
+
+  constructor(
+    line: number,
+    column: number,
+    start: number,
+    end: number,
+    raw: string,
+    virtual: boolean,
+    left: Node,
+    right: Node,
+    isInclusive: boolean
+  ) {
+    super('Range', line, column, start, end, raw, virtual);
+    this.left = left;
+    this.right = right;
+    this.isInclusive = isInclusive;
+  }
+}
+
+export class BinaryOp extends Node {
+  readonly left: Node;
+  readonly right: Node;
+
+  constructor(
+    type: string,
+    line: number,
+    column: number,
+    start: number,
+    end: number,
+    raw: string,
+    virtual: boolean,
+    left: Node,
+    right: Node
+  ) {
+    super(type, line, column, start, end, raw, virtual);
+    this.left = left;
+    this.right = right;
+  }
+}
+
+export class ExtendsOp extends BinaryOp {
+  constructor(
+    line: number,
+    column: number,
+    start: number,
+    end: number,
+    raw: string,
+    virtual: boolean,
+    left: Node,
+    right: Node
+  ) {
+    super('ExtendsOp', line, column, start, end, raw, virtual, left, right);
+  }
+}
+
+export class SeqOp extends BinaryOp {
+  constructor(
+    line: number,
+    column: number,
+    start: number,
+    end: number,
+    raw: string,
+    virtual: boolean,
+    left: Node,
+    right: Node
+  ) {
+    super('SeqOp', line, column, start, end, raw, virtual, left, right);
+  }
+}
+
+export class BaseFunction extends Node {
```

Comment:
I chose to make a `BaseFunction` node type, which should probably be `abstract`, since all the function nodes differ only by `type` and it'll be useful in decaffeinate to be able to say `node instanceof BaseFunction`.